### PR TITLE
Runtime upgrades for ArbOS

### PR DESCRIPTION
**Pull request by edfelten**
This PR includes an apparently working version of runtime upgrades for ArbOS.  It includes an automatic generator that reads the pre-upgrade and post-upgrade .mexe files and automatically generates code to rearrange the old program's globals and install the new program's jump table.  It also makes minor compiler changes, to provide the affordances ArbOS needs to tell whether it's running as an upgrade and to allow remapping of globals. Finally, it adds an alternate startup path that re-initializes just enough to switch over to the new code without destroying any of the old state.

### CODE REVIEW

File: arb_os/accounts.mini [hunk 1]
```diff
@@ -267,23 +267,47 @@ public impure func accountStore_createAccountFromEvmCode(
...
+    let oldAcct = accountStore_get(store, addr);
+    return Some(accountStore_set(
+        store,
+        addr,
+        oldAcct with {
+            contractInfo: Some(struct {
+                code: code,
+                evmJumpTable: evmJumpTable,
+                startCodePoint: initCodePt,
+                storage: (oldAcct.contractInfo?).storage,
```
Comment (edfelten) [inline arb_os/accounts.mini:286]: This should deal with the case where `oldAcct.contractInfo` is None

File: arb_os/main.mini [hunk 2]
```diff
@@ -39,35 +41,52 @@ use std::bytestream::ByteStream;
...
-    txReceipts_init();
-    rollupTracker_init();
-    if (initializePreinstalls() == None<()>) {
-        panic;   // don't try to run without preinstalls and precompiles
+        // now re-do the initializations that set up pointers into the ArbOS code
+        errorHandler_init();
+        evmOps_init();
+        evmCallStack_makeEmpty();   // clear out the old EVM call stack from before the upgrade
+        if (initializePreinstalls(true) == None<()>) {
+            panic;   // don't try to run without preinstalls and precompiles
```
Comment (edfelten) [inline arb_os/main.mini:69]: Add a comment saying that this will abort the upgrade.

File: src/compile/ast.rs [hunk 3]
```diff
@@ -293,18 +330,24 @@ impl Type {
...
-            Type::FixedArray(t, size) => format!("[{}]{}", size, t.display()),
+            Type::Array(t) => format!("[]{}", t.display_indented(indent_level)),
+            Type::FixedArray(t, size) => format!("[{}]{}", size, t.display_indented(indent_level)),
             Type::Struct(fields) => {
                 let mut out = "struct {\n".to_string();
+                for _ in 0..indent_level {
+                    out.push_str("    ");
+                }
                 for field in fields {
                     //This should indent further when dealing with sub-structs
```
Comment (edfelten) [inline src/compile/ast.rs:346]: Remove this comment, because we already do what it is asking us to do.

File: src/compile/mod.rs [hunk 4]
```diff
@@ -174,15 +175,15 @@ impl CompiledProgram {
         int_offset: usize,
         ext_offset: usize,
         func_offset: usize,
-        globals_offset: usize,
+        globals_offset: Vec<GlobalVarDecl>,
```
Comment (edfelten) [inline src/compile/mod.rs:178]: The variable name `globals_offset` is confusing here. Maybe change it to `globals`?

**Review by edfelten — COMMENTED** [review]
One substantive thing to change.

**Review by RGates94 — COMMENTED** [review]
Looks good overall, a few small suggestions.

File: arb_os/accounts.mini (hunk 1 above)
Comment (edfelten) [reply arb_os/accounts.mini]: Fixed.

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**